go/libkb/upak_loader.go

```diff
@@ -673,6 +674,41 @@ func (u *CachedUPAKLoader) lookupUsernameAndDeviceWithInfo(ctx context.Context,
 	return NormalizedUsername(""), "", "", err
 }
 
+func (u *CachedUPAKLoader) CheckDeviceForUIDAndUsername(ctx context.Context, uid keybase1.UID, did keybase1.DeviceID, n NormalizedUsername) (err error) {
+	arg := NewLoadUserByUIDArg(ctx, u.G(), uid).WithForcePoll(true).WithPublicKeyOptional()
+	foundUser := false
+	foundDevice := false
+	isRevoked := false
+	var foundUsername NormalizedUsername
+	u.loadWithInfo(arg, nil, func(upak *keybase1.UserPlusKeysV2AllIncarnations) error {
+		if upak == nil {
+			return nil
+		}
+		foundUser = true
+		foundUsername = NewNormalizedUsername(upak.Current.Username)
+		if pk := upak.FindDevice(did); pk != nil {
+			foundDevice = true
+			if pk.Base.Revocation != nil {
+				isRevoked = true
+			}
+		}
+		return nil
+	}, false)
+	if !foundUser {
+		return UserNotFoundError{UID: uid}
+	}
+	if !foundDevice {
+		return DeviceNotFoundError{Where: "UPAKLoader", ID: did, Loaded: false}
+	}
+	if isRevoked {
+		return NewKeyRevokedError(did.String())
+	}
+	if !n.IsNil() && !foundUsername.Eq(n) {
+		return NewBadUsernameError(fmt.Sprintf("expected %q not %q", foundUsername.String(), n.String()))
```

Comment:
"Bad" username confused me.  I thought BadUsernameError was used for things like `!!pc$`.  Maybe UsernameMismatchError?

Reply:
ok fixed.  `LoggedInWrongUserError`